quarkus-calculator-monolith/src/main/java/com/redhat/training/SolverResource.java

```diff
@@ -0,0 +1,57 @@
+package com.redhat.training;
+
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+import javax.ws.rs.GET;
+import javax.ws.rs.Path;
+import javax.ws.rs.PathParam;
+import javax.ws.rs.Produces;
+import javax.ws.rs.WebApplicationException;
+import javax.ws.rs.core.MediaType;
+import javax.ws.rs.core.Response;
+
+import com.redhat.training.service.SolverService;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class SolverResource implements SolverService {
+    final Logger log = LoggerFactory.getLogger(SolverResource.class);
+
+    static final Pattern multiplyPattern = Pattern.compile("(.*)\\*(.*)");
+    static final Pattern addPattern = Pattern.compile("(.*)\\+(.*)");
+
+    @Override
+    @GET
+    @Path("{equation}")
+    @Produces(MediaType.TEXT_PLAIN)
+    public Float solve(@PathParam("equation") String equation) {
+        log.info("Solving '{}'", equation);
+        try {
+            return Float.valueOf(equation);
+        } catch (NumberFormatException e) {
+            Matcher addMatcher = addPattern.matcher(equation);
+            if (addMatcher.matches()) {
+                return add(addMatcher.group(1),addMatcher.group(2));
+            } 
+            
+            Matcher multiplyMatcher = multiplyPattern.matcher(equation);
+            if (multiplyMatcher.matches()) {
+                return multiply(multiplyMatcher.group(1),multiplyMatcher.group(2));
+            }
+            
+            throw new WebApplicationException(Response.status(Response.Status.BAD_REQUEST).entity("Unable to parse: "+equation).build());
```

Comment:
Could we have the Pattern objects and the this::calculation methods in a Collection and iterate them to avoid the multiple ifs? Maybe since this is for teaching, it is not worth it and it is better to keep it simple.

Pseudo-code (in other words, didn't test or anything, just the idea of a structure):

```

Map<Pattern, BiFunction<String, String, Float>> matchers = new HashMap<>();
matchers.add(addPattern, this::add);
matchers.add(multiplyPattern, this::multiply);
for (var entry : map.entrySet()) {
    var pattern = entry.getKey();
    var function = entry.getValue();
    var matcher = pattern.matcher(equation);
    if (matcher.matches()) {
        return function.apply(matcher.group(1), matcher.group(2));
    }
}
```

Reply:
Yes, we can, but not using a `BiFunction` (note that there are operands that accept a single operator, like `invert` or `sqrt`).
Also, using a `Map` does not ensure the order of the operations. We need to use a sorted collection.

Maybe we can do something like:
```
Map<Pattern, Function<Matcher,Float>> matchers = new LinkedHashMap<>();
matchers.add(addPattern, matcher->add(matcher,group(1),matcher.group(2)));
matchers.add(multiplyPattern, matcher->multiply(matcher,group(1),matcher.group(2)));
...
for (var entry : map.entrySet()) {
    var pattern = entry.getKey();
    var function = entry.getValue();
    var matcher = pattern.matcher(equation);
    if (matcher.matches()) {
        return function.apply(matcher);
    }
}

```

Anyway, I'm not thinking on expanding the calculator much more, so I would leave this as is unless a compelling reason to change it.
KISS.